crates/edr_napi/src/solidity_tests.rs

```diff
@@ -0,0 +1,112 @@
+mod runner;
+mod test_results;
+mod test_suite;
+
+use std::path::Path;
+
+use forge::TestFilter;
+use napi::{
+    threadsafe_function::{
+        ErrorStrategy, ThreadSafeCallContext, ThreadsafeFunction, ThreadsafeFunctionCallMode,
+    },
+    tokio,
+    tokio::runtime,
+    Env, JsFunction,
+};
+use napi_derive::napi;
+
+use crate::solidity_tests::{
+    runner::build_runner, test_results::SuiteResult, test_suite::TestSuite,
+};
+
+/// Executes solidity tests.
+#[napi]
+pub struct SolidityTestRunner {
+    /// The callback to call with the results as they become available.
+    results_callback_fn: ThreadsafeFunction<SuiteResult>,
+}
+
+// The callback has to be passed in the constructor because it's not `Send`.
+#[napi]
+impl SolidityTestRunner {
+    /// Creates a new instance of the SolidityTestRunner.
+    #[doc = "Creates a new instance of the SolidityTestRunner. The callback function will be called with suite results as they finish."]
+    #[napi(constructor)]
+    pub fn new(env: Env, results_callback: JsFunction) -> napi::Result<Self> {
+        let mut results_callback_fn: ThreadsafeFunction<_, ErrorStrategy::CalleeHandled> =
+            results_callback.create_threadsafe_function(
+                // Unbounded queue size
+                0,
+                |ctx: ThreadSafeCallContext<SuiteResult>| Ok(vec![ctx.value]),
+            )?;
+
+        // Allow the event loop to exit before the function is destroyed
+        results_callback_fn.unref(&env)?;
+
+        Ok(Self {
+            results_callback_fn,
+        })
+    }
+
+    #[doc = "Runs the given test suites."]
+    #[napi]
+    pub async fn run_tests(&self, test_suites: Vec<TestSuite>) -> napi::Result<Vec<SuiteResult>> {
+        let test_suites = test_suites
+            .into_iter()
+            .map(|item| Ok((item.id.try_into()?, item.contract.try_into()?)))
+            .collect::<Result<Vec<_>, napi::Error>>()?;
+        let mut runner = build_runner(test_suites)?;
+
+        let (tx_results, mut rx_results) =
+            tokio::sync::mpsc::unbounded_channel::<(String, forge::result::SuiteResult)>();
+        let (tx_end_result, mut rx_end_result) = tokio::sync::mpsc::unbounded_channel();
+
+        let callback_fn = self.results_callback_fn.clone();
+        let runtime = runtime::Handle::current();
+        runtime.spawn(async move {
+            let mut results = Vec::<(String, forge::result::SuiteResult)>::new();
+
+            while let Some(name_and_suite_result) = rx_results.recv().await {
+                results.push(name_and_suite_result.clone());
```

Comment:
If you'd like to avoid the overhead of cloning, you can explore a similar approach to the one I took in other parts of the N-API:

You can wrap the result in an `Arc` and return an `#[napi] struct SomeStruct { inner: Arc<..> }` which implements getter properties to retrieve the contents of the `inner` value.

Reply:
Thanks that's a good idea, but I think it's ok to clone here, because it's not blocking test execution, so I'd keep it as is to keep it simple. A concern could be that the clone is blocking the Tokio worker thread, but I don't think it's big enough to warrant moving it to the blocking thread pool.

Reply:
Ah got it, if there were traces included in the test results, I'd definitely do it, since those are huge. Without traces I think it's not so clear that it's worth it without measuring on a benchmark, so I wouldn't do it in this PR.

Reply:
Sounds good